###
<h2 style="margin-bottom: 0;">Hotel Reviews:</h2>
<h1 style="margin-top: 5px; color: #003580; font-size: 45px; font-weight: bold;">Booking.com</h1>




####
**Import Librerie**



**Pandas:** Manipolazione e analisi dei Dati  
**NumPy:** Calcolo scientifico e matematico  
**Matplotlib:** Creazione di grafici e visualizzazione dati

In [1]:
import pandas as pandas
import numpy as numpy
import matplotlib
import matplotlib.pyplot as plt

print('Pandas:', pandas.__version__)
print('NumPy:', numpy.__version__)
print('Matplotlib:', matplotlib.__version__)

Pandas: 3.0.3
NumPy: 2.4.6
Matplotlib: 3.11.0


## **4.1 Preparazione e Pulizia dei Dati**

(Python + pandas) 

####
**4.1.1 Caricare il file Hotel_Reviews.csv<br>(unico dataset, 515K righe) con Pandas.**

#####
"Hotel_Reviews.csv"

In [2]:
df = pandas.read_csv("Hotel_Reviews.csv")

#####
4.1.1.1 Copia profonda di sicurezza

df.copy() = df.copy(deep=True)<br> è l'impostazione prefedifinita del comando<br>deep=True --> copia profonda

In [3]:
df_original = df.copy(deep=True)

#####
4.1.1.2 DataFrame: .head() e .tail()

In [4]:
df_original.head(10)

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
0,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,8/3/2017,7.7,Hotel Arena,Russia,I am so angry that i made this post available...,397,1403,Only the park outside of the hotel was beauti...,11,7,2.9,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968
1,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,8/3/2017,7.7,Hotel Arena,Ireland,No Negative,0,1403,No real complaints the hotel was great great ...,105,7,7.5,"[' Leisure trip ', ' Couple ', ' Duplex Double...",0 days,52.360576,4.915968
2,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/31/2017,7.7,Hotel Arena,Australia,Rooms are nice but for elderly a bit difficul...,42,1403,Location was good and staff were ok It is cut...,21,9,7.1,"[' Leisure trip ', ' Family with young childre...",3 days,52.360576,4.915968
3,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/31/2017,7.7,Hotel Arena,United Kingdom,My room was dirty and I was afraid to walk ba...,210,1403,Great location in nice surroundings the bar a...,26,1,3.8,"[' Leisure trip ', ' Solo traveler ', ' Duplex...",3 days,52.360576,4.915968
4,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/24/2017,7.7,Hotel Arena,New Zealand,You When I booked with your company on line y...,140,1403,Amazing location and building Romantic setting,8,3,6.7,"[' Leisure trip ', ' Couple ', ' Suite ', ' St...",10 days,52.360576,4.915968
5,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/24/2017,7.7,Hotel Arena,Poland,Backyard of the hotel is total mess shouldn t...,17,1403,Good restaurant with modern design great chil...,20,1,6.7,"[' Leisure trip ', ' Group ', ' Duplex Double ...",10 days,52.360576,4.915968
6,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/17/2017,7.7,Hotel Arena,United Kingdom,Cleaner did not change our sheet and duvet ev...,33,1403,The room is spacious and bright The hotel is ...,18,6,4.6,"[' Leisure trip ', ' Group ', ' Duplex Twin Ro...",17 days,52.360576,4.915968
7,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/17/2017,7.7,Hotel Arena,United Kingdom,Apart from the price for the brekfast Everyth...,11,1403,Good location Set in a lovely park friendly s...,19,1,10.0,"[' Leisure trip ', ' Couple ', ' Duplex Double...",17 days,52.360576,4.915968
8,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/9/2017,7.7,Hotel Arena,Belgium,Even though the pictures show very clean room...,34,1403,No Positive,0,3,6.5,"[' Leisure trip ', ' Couple ', ' Duplex Double...",25 days,52.360576,4.915968
9,s Gravesandestraat 55 Oost 1092 AA Amsterdam ...,194,7/8/2017,7.7,Hotel Arena,Norway,The aircondition makes so much noise and its ...,15,1403,The room was big enough and the bed is good T...,50,1,7.9,"[' Leisure trip ', ' Couple ', ' Large King Ro...",26 days,52.360576,4.915968


In [5]:
df_original.tail(10)

,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
515728,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,9/27/2015,8.1,Atlantis Hotel Vienna,Qatar,Everything,2,2823,Nothing,2,1,4.2,"[' Leisure trip ', ' Couple ', ' Standard Doub...",676 day,48.203745,16.335677
515729,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,9/26/2015,8.1,Atlantis Hotel Vienna,Brazil,No Negative,0,2823,The breakfast was excelent lots of fress frui...,27,7,10.0,"[' Leisure trip ', ' Couple ', ' Standard Doub...",677 day,48.203745,16.335677
515730,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,9/20/2015,8.1,Atlantis Hotel Vienna,Israel,no free parking the room was in another build...,20,2823,large room,3,2,6.7,"[' Leisure trip ', ' Family with young childre...",683 day,48.203745,16.335677
515731,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,9/7/2015,8.1,Atlantis Hotel Vienna,Germany,No parking Public parking garage is 15 Euro p...,12,2823,Good breakfast,3,1,7.5,"[' Business trip ', ' Solo traveler ', ' Stand...",696 day,48.203745,16.335677
515732,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,9/6/2015,8.1,Atlantis Hotel Vienna,Kuwait,No Negative,0,2823,helpful staff allowed me to check in early as...,18,6,10.0,"[' Leisure trip ', ' Family with young childre...",697 day,48.203745,16.335677
515733,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/30/2015,8.1,Atlantis Hotel Vienna,Kuwait,no trolly or staff to help you take the lugga...,14,2823,location,2,8,7.0,"[' Leisure trip ', ' Family with older childre...",704 day,48.203745,16.335677
515734,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/22/2015,8.1,Atlantis Hotel Vienna,Estonia,The hotel looks like 3 but surely not 4,11,2823,Breakfast was ok and we got earlier check in,11,12,5.8,"[' Leisure trip ', ' Family with young childre...",712 day,48.203745,16.335677
515735,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/19/2015,8.1,Atlantis Hotel Vienna,Egypt,The ac was useless It was a hot week in vienn...,19,2823,No Positive,0,3,2.5,"[' Leisure trip ', ' Family with older childre...",715 day,48.203745,16.335677
515736,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/17/2015,8.1,Atlantis Hotel Vienna,Mexico,No Negative,0,2823,The rooms are enormous and really comfortable...,25,3,8.8,"[' Leisure trip ', ' Group ', ' Standard Tripl...",717 day,48.203745,16.335677
515737,Wurzbachgasse 21 15 Rudolfsheim F nfhaus 1150 ...,168,8/9/2015,8.1,Atlantis Hotel Vienna,Hungary,I was in 3rd floor It didn t work Free Wife,13,2823,staff was very kind,6,1,8.3,"[' Leisure trip ', ' Family with young childre...",725 day,48.203745,16.335677


EDA - Analisi esplorativa iniziale

####
**4.1.2 Esplorare il dataframe:<br>dimensioni, tipi di dato, valori mancanti, distribuzione Reviewer_Score e Average_Score.**

#####
4.1.2.1 Dimensioni

In [6]:
df_original.shape

(515738, 17)

Analisi del Dataset:<br>
È composto da 515738 righe e 17 colonne

Le Colonne del Dataset sono:

In [7]:
df_original.columns

Index(['Hotel_Address', 'Additional_Number_of_Scoring', 'Review_Date',
       'Average_Score', 'Hotel_Name', 'Reviewer_Nationality',
       'Negative_Review', 'Review_Total_Negative_Word_Counts',
       'Total_Number_of_Reviews', 'Positive_Review',
       'Review_Total_Positive_Word_Counts',
       'Total_Number_of_Reviews_Reviewer_Has_Given', 'Reviewer_Score', 'Tags',
       'days_since_review', 'lat', 'lng'],
      dtype='str')

Analisi delle Colonne:<br>
17 colonne che possono essere raggruppate in tre macro-categorie logiche:

1. **Informazioni sull'Hotel:**
   * `Hotel_Name` e `Hotel_Address`: Nome e indirizzo della struttura.
   * `Average_Score` e `Total_Number_Of_Reviews`: Il punteggio medio storico dell'hotel su Booking e il numero totale di recensioni ricevute.
   * `lat` e `lng`: Coordinate geografiche (latitudine e longitudine).

2. **Informazioni su Reviewer:**
   * `Reviewer_Nationality`: Nazionalità del viaggiatore.
   * `Total_Number_Of_Reviews_Reviewer_Has_Given`: Quante recensioni ha scritto in totale questo utente in passato.

3. **Dettagli della Recensione:**
   * `Review_Date`: Data in cui è stata lasciata la recensione.
   * `Reviewer_Score`: Il voto specifico dato dal singolo utente a quel soggiorno.
   * `Negative_Review` e `Positive_Review`: I testi descrittivi degli aspetti negativi e positivi.
   * `Review_Total_Negative_Word_Counts` e `Review_Total_Positive_Word_Counts`: Il conteggio delle parole usate nei commenti.
   * `Tags`: I tag associati al viaggio (es. tipo di stanza, durata del soggiorno, tipo di gruppo).
   * `days_since_review`: Il tempo trascorso tra la recensione e la data di estrazione dei dati.

#####
4.1.2.2 Tipi di dato<br>
(Numero di righe/colonne, dtype di ogni colonna, non-null)

In [8]:
df_original.info()

<class 'pandas.DataFrame'>
RangeIndex: 515738 entries, 0 to 515737
Data columns (total 17 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   Hotel_Address                               515738 non-null  str    
 1   Additional_Number_of_Scoring                515738 non-null  int64  
 2   Review_Date                                 515738 non-null  str    
 3   Average_Score                               515738 non-null  float64
 4   Hotel_Name                                  515738 non-null  str    
 5   Reviewer_Nationality                        515738 non-null  str    
 6   Negative_Review                             515738 non-null  str    
 7   Review_Total_Negative_Word_Counts           515738 non-null  int64  
 8   Total_Number_of_Reviews                     515738 non-null  int64  
 9   Positive_Review                             515738 non-null  str    
 10  Review_

Analisi dei tipi di dato:

* Le colonne numeriche sono suddivise in:
  * **int64** (conteggi interi)
  * **float64** (punteggi decimali e coordinate geografiche)
* I dati testuali (come i nomi degli hotel e le recensioni vere e proprie) sono identificati come stringhe (**str**).
* **Nota sui valori mancanti:**<br> Le colonne `lat` (latitudine) e `lng` (longitudine) presentano meno valori rispetto al totale<br> (**512.470** non-null invece di 515.738).<br> Questo evidenzia la presenza di coordinate mancanti (`NaN`) per alcuni hotel.
* La colonna `Review_Date` viene attualmente letta come testo (`str`) e potrebbe essere utile convertirla in formato data (`datetime`) per facilitarne analisi temporali.

#####
4.1.2.3 Valori mancanti (NaN / Null)

In [9]:
df_original.isnull().sum()[df.isnull().sum() > 0]

lat    3268
lng    3268
dtype: int64

In [10]:
# Numero assoluto di valori mancanti
valori_mancanti = df_original.isnull().sum()[df_original.isnull().sum() > 0]

# Percentuale rispetto al totale del dataframe
percentuale_mancanti = (valori_mancanti / len(df_original)) * 100

# Uniamo i dati
analisi_mancanti = pandas.DataFrame({
    'Valori Mancanti': valori_mancanti,
    'Percentuale (%)': percentuale_mancanti.round(2)
})

# Tabella dei missing values
analisi_mancanti

,Valori Mancanti,Percentuale (%)
lat,3268,0.63
lng,3268,0.63


Analisi dei Valori mancanti:
1. **Entità del problema:** Le uniche colonne che presentano dati mancanti sono `lat` (latitudine) e `lng` (longitudine), con esattamente **3.268 righe vuote** ciascuna.
2. **Impatto percentuale:** Questi dati mancanti rappresentano appena lo **0,63%** dell'intero dataset (che conta oltre 515.000 righe) e si tratta di una quota estremamente ridotta.
3. **Correlazione:** Il fatto che il numero di `NaN` sia identico per entrambe le colonne suggerisce che il problema si verifichi in contemporanea: per quegli specifici hotel manca del tutto la geolocalizzazione.

**Gestione "missing values":**<br>
Dato che la percentuale è inferiore all'1%, in fase di pulizia dei dati si possono valutare tre strade possibili:
* **Rimozione:** Eliminare le 3.268 righe (l'impatto sul dataset complessivo sarebbe trascurabile).
* **Imputazione manuale:** Identificare quali hotel non hanno le coordinate e inserire latitudine e longitudine cercandole su Google Maps/OpenStreetMap.
* **Conservazione:** Lasciare i valori vuoti se non si utilizzeranno le coordinate geografiche.

#####
4.1.2.4 Distribuzione Reviewer_Score e Average_Score.

In [11]:
df_original[['Reviewer_Score', 'Average_Score']].describe()

,Reviewer_Score,Average_Score
count,515738.000000,515738.000000
mean,8.395077,8.397487
std,1.637856,0.548048
min,2.500000,5.200000
25%,7.500000,8.100000
50%,8.800000,8.400000
75%,9.600000,8.800000
max,10.000000,9.800000


In [12]:
df_original[['Reviewer_Score', 'Average_Score']].describe().round(2)

,Reviewer_Score,Average_Score
count,515738.00,515738.00
mean,8.40,8.40
std,1.64,0.55
min,2.50,5.20
25%,7.50,8.10
50%,8.80,8.40
75%,9.60,8.80
max,10.00,9.80


Analisi della Distribuzione dei Punteggi:

Il confronto tra le metriche statistiche di `Reviewer_Score` (voto del singolo cliente) e `Average_Score` (media globale dell'hotel) evidenzia dei comportamenti molto interessanti:

1. **La Media dei Voti (`mean`):**
   * La media generale è quasi identica per entrambe le variabili (**~8.39**).<br> Questo ci dice che in media gli hotel europei del dataset hanno una valutazione complessiva molto alta, posizionandosi nella fascia dell'eccellenza.

2. **La Volatilità e Deviazione Standard (`std`):**
   * `Reviewer_Score` ha una deviazione standard molto più alta (**1.64**) rispetto a `Average_Score` (**0.55**). 
   * Questo conferma che i voti dei singoli utenti sono molto più "ballerini" ed emotivi, spaziando facilmente da stroncature a punteggi massimi,<br> mentre la media dell'hotel tende a stabilizzarsi e a variare molto meno.

3. **I Valori Estremi (`min` e `max`):**
   * **Minimi:** Il voto più basso mai dato da un singolo cliente è **2.5**, mentre l'hotel peggiore del dataset ha comunque una media del **5.2**<br> (quindi nessun hotel è considerato un disastro totale nel lungo periodo).
   * **Massimi:** Molti utenti hanno assegnato il punteggio massimo di **10.0**, mentre la media globale di hotel più alta nel dataset si ferma a **9.8**.

4. **I Percentili (Concentrazione dei voti):**
   * Per il `Reviewer_Score`, il 50% dei voti (`50%` o mediana) è pari a **8.8** e ben il 25% delle recensioni (`75%`) ha un voto da **9.6** in su. 
   * Questo dimostra un chiaro **bias positivo**:<br> gli utenti di Booking.com tendono a lasciare recensioni molto generose e si concentrano massicciamente verso i voti più alti, lasciando i voti sotto la sufficienza solo a rari casi eccezionali.

DATA CLEANING

###
**4.1.3 Pulizia di base**


* Gestione missing values.
* Conversione in datetime ed estrazione anno/mese.
* Rimozione duplicati esatti.
* Creazione variabili derivate.

*(Nota: il missing values di `Reviewer_Nationality` è completato in 4.1.3.4.3, dopo la creazione di `City`, necessaria per verificarne il pattern)*

#### 
<span style="color:red">**4.1.3.0 Pulizia degli spazi nelle colonne testuali**</span>



Alcune colonne testuali contengono spazi bianchi extra all'inizio/fine dei valori (es. " United Kingdom " invece di "United Kingdom")<br>scoperto durante le query SQL e le analisi del notebook 02.<br> 
Si effettua la pulizia alla fonte, così il DATASET.csv salvato è già corretto per tutti gli usi successivi (database e analisi).

In [13]:
# Rimozione spazi bianchi iniziali/finali dalle colonne testuali principali
df['Reviewer_Nationality'] = df['Reviewer_Nationality'].str.strip()
df['Hotel_Name'] = df['Hotel_Name'].str.strip()
df['Hotel_Address'] = df['Hotel_Address'].str.strip()

# Verifica su un caso noto: 'United Kingdom' deve avere esattamente 14 caratteri
# In precedenza 16 caratteri = ' United Kingdom '
uk_len = df[df['Reviewer_Nationality'].str.contains('Kingdom')]['Reviewer_Nationality'].str.len().unique()
print(f"Lunghezza 'United Kingdom' dopo strip: {uk_len} (attesa: 14)")

Lunghezza 'United Kingdom' dopo strip: [14] (attesa: 14)


####
4.1.3.1 Gestione missing values


Analisi dei valori mancanti effettuata in precedenza:

"**Gestione "missing values":**<br>
Dato che la percentuale è inferiore all'1%, in fase di pulizia dei dati si possono valutare tre strade possibili:
* **Rimozione:** Eliminare le 3.268 righe (l'impatto sul dataset complessivo sarebbe trascurabile).
* **Imputazione manuale:** Identificare quali hotel non hanno le coordinate e inserire latitudine e longitudine cercandole su Google Maps/OpenStreetMap.
* **Conservazione:** Lasciare i valori vuoti se non si utilizzeranno le coordinate geografiche."

Prima di procedere alla rimozione dei record contenenti valori mancanti (NaN) nelle colonne 'lat' e 'lng',<br>è fondamentale verificare se l'assenza di tali dati sia puramente casuale oppure se risponda a un pattern sistematico.

Se i valori mancanti fossero concentrati interamente su specifici hotel, la loro cancellazione introdurrebbe un bias geografico,<br> eliminando completamente quelle strutture dalle successive analisi.

In [14]:
# Numero totale di hotel unici nel dataset
# Usiamo la chiave composta (Hotel_Name + Hotel_Address) invece del solo nome,
# perché un nome può essere condiviso da hotel diversi in città diverse
# (es. "Hotel Regina" esiste a Barcelona, Vienna e Milano: 3 hotel indipendenti,
# non un duplicato dello stesso hotel)

totale_hotel_per_nome = df_original['Hotel_Name'].nunique()
totale_hotel_reali = df_original.drop_duplicates(subset=['Hotel_Name', 'Hotel_Address']).shape[0]

print(f"Nomi hotel unici (solo Hotel_Name): {totale_hotel_per_nome}")
print(f"Hotel realmente distinti (Hotel_Name + Hotel_Address): {totale_hotel_reali}")

Nomi hotel unici (solo Hotel_Name): 1492
Hotel realmente distinti (Hotel_Name + Hotel_Address): 1494


In [15]:
# Hotel coinvolti dai valori mancanti
righe_senza_geo = df_original[df_original['lat'].isnull()]
elenco_hotel_mancanti = righe_senza_geo['Hotel_Name'].value_counts().reset_index()
elenco_hotel_mancanti.index = elenco_hotel_mancanti.index + 1

#Risultato finale con il conteggio numerico 
elenco_hotel_mancanti

,Hotel_Name,count
1,Fleming s Selection Hotel Wien City,658
2,Hotel City Central,563
3,Hotel Atlanta,389
4,Maison Albar Hotel Paris Op ra Diamond,290
5,Hotel Daniel Vienna,245
6,Hotel Pension Baron am Schottentor,223
7,Austria Trend Hotel Schloss Wilhelminenberg Wien,194
8,Derag Livinghotel Kaiser Franz Joseph Vienna,147
9,NH Collection Barcelona Podium,146
10,City Hotel Deutschmeister,93


In [16]:
# Hotel coinvolti dai valori mancanti e verifica delle Coordinate_Presenti, Coordinante_Mancanti e Recensioni_Totali.
nomi_hotel_mancanti = df_original[df_original['lat'].isnull()]['Hotel_Name'].unique()

df_filtrato = df_original[df_original['Hotel_Name'].isin(nomi_hotel_mancanti)]

conteggio_geo = df_filtrato.groupby('Hotel_Name').agg(
    Coordinate_Presenti=('lat', lambda x: x.notnull().sum()),
    Coordinate_Mancanti=('lat', lambda x: x.isnull().sum()),
    Recensioni_Totali=('Hotel_Name', 'count')
).reset_index()

conteggio_geo = conteggio_geo.sort_values(by='Coordinate_Mancanti', ascending=False).reset_index(drop=True)

conteggio_geo.index = conteggio_geo.index + 1

#Tabella finale
conteggio_geo

,Hotel_Name,Coordinate_Presenti,Coordinate_Mancanti,Recensioni_Totali
1,Fleming s Selection Hotel Wien City,0,658,658
2,Hotel City Central,0,563,563
3,Hotel Atlanta,0,389,389
4,Maison Albar Hotel Paris Op ra Diamond,0,290,290
5,Hotel Daniel Vienna,0,245,245
6,Hotel Pension Baron am Schottentor,0,223,223
7,Austria Trend Hotel Schloss Wilhelminenberg Wien,0,194,194
8,Derag Livinghotel Kaiser Franz Joseph Vienna,0,147,147
9,NH Collection Barcelona Podium,0,146,146
10,City Hotel Deutschmeister,0,93,93


Analisi dei Valori mancanti e del Bias Geografico:

Dall'analisi esplorativa iniziale è emersa l'assenza di coordinate geografiche (`lat` e `lng`) per **3.268 record**,<br>pari a circa lo **0,63%** dell'intero dataset (515K righe). 

Prima di procedere a una rimozione tramite `.dropna()`, è stata effettuata una verifica approfondita per capire la natura di questi dati mancanti.
L'obiettivo era determinare se i dati mancassero in modo del tutto casuale o secondo un pattern sistematico legato a specifiche strutture.

L'analisi aggregata ha rivelato una situazione molto chiara e critica:
* **Nomi hotel nel dataset:** Il dataset contiene **1.492 nomi hotel distinti** (attenzione: non equivale al numero di hotel realmente distinti; un caso emerso più avanti nel progetto, "Hotel Regina", è un nome condiviso da 3 hotel indipendenti in città diverse e il conteggio reale di hotel univoci, verificato nel database, è 1.494).
* **I valori mancanti non sono casuali:** I 3.268 valori mancanti sono concentrati interamente ed esclusivamente su **appena 17 hotel**<br>(l'1,1% delle strutture totali).
* **Mancanza totale:** Per ciascuno di questi 17 hotel, la colonna `Coordinate_Presenti` è pari a **0**.<br>
Questo significa che per queste specifiche strutture non esiste alcuna coordinata geografica in nessuna delle loro recensioni nel dataset originale.


**Conclusione:**<br>
Eliminare queste righe con un semplice `.dropna()` introdurrebbe un **bias geografico sistematico**, poiché significherebbe cancellare
non un campione casuale di recensioni, ma tutte le recensioni relative a 17 hotel specifici: si eliminerebbero così
sistematicamente quelle strutture (e le città a cui appartengono) dall'analisi, invece di perdere un dato isolato e ininfluente.

<span style="color:red">Imputazione manuale delle coordinate mancanti</span>

In [17]:
# Coordinate reali recuperate per i 17 hotel 
coordinate_mancanti = {
    "Fleming s Selection Hotel Wien City": (48.209402, 16.353446),
    "Hotel City Central": (48.213587, 16.379908),
    "Hotel Atlanta": (48.220410, 16.355819),
    "Maison Albar Hotel Paris Op ra Diamond": (48.875342, 2.323393),
    "Hotel Daniel Vienna": (48.188815, 16.383779),
    "Hotel Pension Baron am Schottentor": (48.216800, 16.359906),
    "Austria Trend Hotel Schloss Wilhelminenberg Wien": (48.219569, 16.285592),
    "Derag Livinghotel Kaiser Franz Joseph Vienna": (48.246019, 16.341254),
    "NH Collection Barcelona Podium": (41.391548, 2.177894),
    "City Hotel Deutschmeister": (48.220852, 16.366611),
    "Hotel Park Villa": (48.233577, 16.345672),
    "Cordial Theaterhotel Wien": (48.209549, 16.351463),
    "Holiday Inn Paris Montmartre": (48.888909, 2.333153),
    "Roomz Vienna": (48.222761, 16.393487),
    "Mercure Paris Gare Montparnasse": (48.839996, 2.323562),
    "Renaissance Barcelona Hotel": (41.392685, 2.167398),
    "Hotel Advance": (41.383239, 2.162950),
}

# Coordinate su df (df_original resta invariato)
for hotel, (lat, lng) in coordinate_mancanti.items():
    mask = df['Hotel_Name'] == hotel
    df.loc[mask, 'lat'] = lat
    df.loc[mask, 'lng'] = lng

# Verifica
df[['lat', 'lng']].isnull().sum()

lat    0
lng    0
dtype: int64

In [18]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 515738 entries, 0 to 515737
Data columns (total 17 columns):
 #   Column                                      Non-Null Count   Dtype  
---  ------                                      --------------   -----  
 0   Hotel_Address                               515738 non-null  str    
 1   Additional_Number_of_Scoring                515738 non-null  int64  
 2   Review_Date                                 515738 non-null  str    
 3   Average_Score                               515738 non-null  float64
 4   Hotel_Name                                  515738 non-null  str    
 5   Reviewer_Nationality                        515738 non-null  str    
 6   Negative_Review                             515738 non-null  str    
 7   Review_Total_Negative_Word_Counts           515738 non-null  int64  
 8   Total_Number_of_Reviews                     515738 non-null  int64  
 9   Positive_Review                             515738 non-null  str    
 10  Review_

####
4.1.3.2 Conversione in datetime ed estrazione anno/mese.

In [19]:
# 1. Prime righe della colonna Review_Date
print("Prime righe di Review_Date:")
print(df_original['Review_Date'].head())

print("\n" + "="*40 + "\n")

# 2. Tipo di dato attuale della colonna
print(df_original['Review_Date'].dtype)

Prime righe di Review_Date:
0     8/3/2017
1     8/3/2017
2    7/31/2017
3    7/31/2017
4    7/24/2017
Name: Review_Date, dtype: str


str


In [20]:
# 1.Colonna in datetime specificando il formato corretto (Mese/Giorno/Anno)
df['Review_Date'] = pandas.to_datetime(df['Review_Date'], format='%m/%d/%Y')

# 2. Estrazione di anno e mese con creazione delle due nuove colonne
df['Review_Year'] = df['Review_Date'].dt.year
df['Review_Month'] = df['Review_Date'].dt.month

# 3. VErifica del risultato
df[['Review_Date', 'Review_Year', 'Review_Month']].head()

,Review_Date,Review_Year,Review_Month
0,2017-08-03,2017,8
1,2017-08-03,2017,8
2,2017-07-31,2017,7
3,2017-07-31,2017,7
4,2017-07-24,2017,7


La colonna `Review_Date` è stata originariamente caricata come stringa (`str`) seguendo la formattazione anglosassone (Mese/Giorno/Anno). 

Per abilitare le analisi temporali richieste dalle specifiche del progetto, la colonna viene convertita nel tipo di dato nativo `datetime` applicando il pattern esplicito `format='%m/%d/%Y'`.<br> Successivamente, vengono estratte due nuove variabili numeriche d'appoggio per raggruppare i trend:
* **`Review_Year`**: Anno della recensione.
* **`Review_Month`**: Mese della recensione.

In [21]:
# 1. Prime righe della colonna Review_Date
print("Prime righe di Review_Date:")
print(df['Review_Date'].head())

print("\n" + "="*40 + "\n")

# 2. Tipo di dato attuale della colonna
print(df['Review_Date'].dtype)

Prime righe di Review_Date:
0   2017-08-03
1   2017-08-03
2   2017-07-31
3   2017-07-31
4   2017-07-24
Name: Review_Date, dtype: datetime64[us]


datetime64[us]


####
4.1.3.3 Rimozione duplicati esatti.



In questa fase si identificano e vengono rimosse le righe completamente duplicate all'interno del dataset di lavoro, per evitare di alterare i conteggi complessivi e falsare le analisi statistiche successive.

In [22]:
# 1. Righe duplicate esatte presenti nel dataframe
valori_duplicati = df.duplicated().sum()
print(f"Numero di righe completamente duplicate: {valori_duplicati}")

Numero di righe completamente duplicate: 526


In [23]:
# 2. Tutte le righe coinvolte in un duplicato (sia la prima occorrenza che le copie)
# keep=False marca come duplicate TUTTE le occorrenze coinvolte, non solo le "copie in eccesso"
righe_duplicate = df_original[df_original.duplicated(keep=False)]

# 3. Ordine per Hotel_Name e Review_Date così le coppie di duplicati compaiono vicine
righe_duplicate = righe_duplicate.sort_values(by=['Hotel_Name', 'Review_Date'])

print(f"Totale righe coinvolte in duplicati (originali + copie): {len(righe_duplicate)}")
righe_duplicate.head(10)

Totale righe coinvolte in duplicati (originali + copie): 1052


,Hotel_Address,Additional_Number_of_Scoring,Review_Date,Average_Score,Hotel_Name,Reviewer_Nationality,Negative_Review,Review_Total_Negative_Word_Counts,Total_Number_of_Reviews,Positive_Review,Review_Total_Positive_Word_Counts,Total_Number_of_Reviews_Reviewer_Has_Given,Reviewer_Score,Tags,days_since_review,lat,lng
444764,Sch nbrunner Stra e 41 05 Margareten 1050 Vien...,144,2/23/2016,8.4,Arthotel ANA Prime,China,Price only includes room Everything else is e...,34,1813,Good location very clean and modern nice staff,10,26,7.5,"[' Leisure trip ', ' Couple ', ' Standard Doub...",527 day,48.192558,16.357607
444765,Sch nbrunner Stra e 41 05 Margareten 1050 Vien...,144,2/23/2016,8.4,Arthotel ANA Prime,China,Price only includes room Everything else is e...,34,1813,Good location very clean and modern nice staff,10,26,7.5,"[' Leisure trip ', ' Couple ', ' Standard Doub...",527 day,48.192558,16.357607
443175,Scarsdale Place Kensington Kensington and Chel...,1831,11/8/2016,8.1,Copthorne Tara Hotel London Kensington,United Kingdom,No Negative,0,7105,Location Breakfast,3,11,7.9,"[' Business trip ', ' Couple ', ' Standard Dou...",268 day,51.499046,-0.191707
443176,Scarsdale Place Kensington Kensington and Chel...,1831,11/8/2016,8.1,Copthorne Tara Hotel London Kensington,United Kingdom,No Negative,0,7105,Location Breakfast,3,11,7.9,"[' Business trip ', ' Couple ', ' Standard Dou...",268 day,51.499046,-0.191707
445413,Schepenbergweg 50 Zuidoost 1105 AT Amsterdam N...,315,8/31/2016,7.8,Fletcher Hotel Amsterdam,United Kingdom,No Negative,0,3773,Very nice hotel,5,6,10.0,"[' Leisure trip ', ' Couple ', ' Deluxe Double...",337 day,52.292457,4.944434
445414,Schepenbergweg 50 Zuidoost 1105 AT Amsterdam N...,315,8/31/2016,7.8,Fletcher Hotel Amsterdam,United Kingdom,No Negative,0,3773,Very nice hotel,5,6,10.0,"[' Leisure trip ', ' Couple ', ' Deluxe Double...",337 day,52.292457,4.944434
239296,7 rue Casimir Delavigne 6th arr 75006 Paris Fr...,44,1/22/2016,9.0,H tel Baume,Australia,No Negative,0,253,This is a beautiful 1930 s boutique hotel cle...,48,8,10.0,"[' Leisure trip ', ' Couple ', ' The Duplex ',...",559 day,48.850152,2.339278
239384,7 rue Casimir Delavigne 6th arr 75006 Paris Fr...,44,1/22/2016,9.0,H tel Baume,Australia,No Negative,0,253,This is a beautiful 1930 s boutique hotel cle...,48,8,10.0,"[' Leisure trip ', ' Couple ', ' The Duplex ',...",559 day,48.850152,2.339278
239277,7 rue Casimir Delavigne 6th arr 75006 Paris Fr...,44,1/22/2017,9.0,H tel Baume,Belgium,breakfast was good but just that at 6th floor...,30,253,the position excellent the staff charming and...,22,9,7.9,"[' Leisure trip ', ' Couple ', ' Classic Doubl...",193 day,48.850152,2.339278
239365,7 rue Casimir Delavigne 6th arr 75006 Paris Fr...,44,1/22/2017,9.0,H tel Baume,Belgium,breakfast was good but just that at 6th floor...,30,253,the position excellent the staff charming and...,22,9,7.9,"[' Leisure trip ', ' Couple ', ' Classic Doubl...",193 day,48.850152,2.339278


In [24]:
# 4. Rimozione effettiva dei duplicati esatti solo dal dataframe di lavoro df
righe_prima = len(df)
df = df.drop_duplicates()
righe_dopo = len(df)

# 5. Controllo e riepilogo
print(f"Righe in df prima della rimozione: {righe_prima}")
print(f"Righe in df dopo la rimozione:    {righe_dopo}")
print(f"Duplicati effettivi rimossi:      {righe_prima - righe_dopo}")

Righe in df prima della rimozione: 515738
Righe in df dopo la rimozione:    515212
Duplicati effettivi rimossi:      526


In [25]:
# Controllo
duplicati = df.duplicated().sum()
print(f"Righe duplicate trovate: {duplicati}")

Righe duplicate trovate: 0


#### 
4.1.3.4 Creazione variabili derivate



Costruzione delle colonne aggiuntive richieste dalle istruzioni per le analisi e i grafici delle fasi successive.

#####
4.1.3.4.1 Total_Review_Length


Richiesta al punto 4.3.5 delle istruzioni<br>
("Distribuzione Review_Length per categoria score", "Correlazione lunghezza recensione vs Reviewer_Score")<br>
e per i grafici 5 e 6<br>
(istogramma e scatter plot sulla lunghezza recensione).

In [26]:
# Lunghezza totale della recensione = parole della recensione positiva + parole della recensione negativa
df['Total_Review_Length'] = (
    df['Review_Total_Positive_Word_Counts'] + df['Review_Total_Negative_Word_Counts']
)

# Verifica
df[['Review_Total_Positive_Word_Counts', 'Review_Total_Negative_Word_Counts', 'Total_Review_Length']].head()

,Review_Total_Positive_Word_Counts,Review_Total_Negative_Word_Counts,Total_Review_Length
0,11,397,408
1,105,0,105
2,21,42,63
3,26,210,236
4,8,140,148


##### 
4.1.3.4.2 Score_Category

Serve per il grafico 5 obbligatorio<br>("Histogram Review Length per categoria score", punto 4.5.5)<br>e per la query SQL "Distribuzione Review_Length per categoria score" (punto 4.3.5).<br> Reviewer_Score (0-10) suddiviso in 4 fasce discrete.

In [27]:
# Reviewer_Score in 4 fasce

# Aggiornamento:
# right=False rende gli intervalli chiusi a sinistra [0,5), [5,7), [7,9), [9,10.1)
# per allinearsi esattamente ai confini usati nella query 
# SQL 2.10 Distribuzione lunghezza recensione per categoria di score (WHEN score < 5, < 7, < 9)
# Nota: l'ultimo bin arriva a 10.1 perché con right=False l'intervallo [9,10) escluderebbe
# il valore massimo 10.0, che deve invece ricadere in "Ottimo"

df['Score_Category'] = pandas.cut(
    df['Reviewer_Score'],
    bins=[0, 5, 7, 9, 10.1],
    labels=['Basso (0-5)', 'Medio (5-7)', 'Buono (7-9)', 'Ottimo (9-10)'],
    right=False
)

# Verifica

# Aggiornamento
# Verifica: i conteggi devono coincidere con la query SQL 2.10
# (attesi: Basso=22232, Medio=64485, Buono=181668, Ottimo=246827)
print(df['Score_Category'].value_counts().sort_index())

Score_Category
Basso (0-5)       22232
Medio (5-7)       64485
Buono (7-9)      181668
Ottimo (9-10)    246827
Name: count, dtype: int64


Analisi:

Quasi metà delle recensioni (246.827) sono "Ottimo",<br>
solo il 4,3% circa (22.232 su 515.212) è "Basso"

##### 
4.1.3.4.3 City

Necessaria per tutte le analisi geografiche richieste:<br>"Distribuzione recensioni per città", "Score medio per nazione, top città" (punto 4.3.4),<br> più i grafici 3 e 7 (bar chart top città, score medio per città).<br>Il dataset non fornisce una colonna città pronta: va ricavata da Hotel_Address.

In [28]:
# Estrazione dell'ultima parola di ogni indirizzo per verificare il nome dei paesi presenti
paesi = df['Hotel_Address'].apply(lambda x: x.split()[-1])
print(f"Numero di paesi unici: {paesi.nunique()}")
print(f"Paesi trovati: {sorted(paesi.unique())}")

Numero di paesi unici: 6
Paesi trovati: ['Austria', 'France', 'Italy', 'Kingdom', 'Netherlands', 'Spain']


In [29]:
# Estrazione paese e "città candidata" (penultima parola) per ogni riga
df_check = df[['Hotel_Address']].copy()
df_check['Paese'] = df_check['Hotel_Address'].apply(lambda x: x.split()[-1])
df_check['Citta_Candidata'] = df_check['Hotel_Address'].apply(lambda x: x.split()[-2])

# Tabella riassuntiva: per ogni paese, quali città candidate uniche si trovano e quante righe coinvolgono
riepilogo = df_check.groupby('Paese')['Citta_Candidata'].agg(['unique', 'nunique'])
riepilogo.columns = ['Citta_Trovate', 'Numero_Citta_Uniche']

# Anomalia caso United Kingdom
riepilogo['Caso_Anomalo'] = riepilogo.index == 'Kingdom'

print(riepilogo)

            Citta_Trovate  Numero_Citta_Uniche  Caso_Anomalo
Paese                                                       
Austria          [Vienna]                    1         False
France            [Paris]                    1         False
Italy             [Milan]                    1         False
Kingdom          [United]                    1          True
Netherlands   [Amsterdam]                    1         False
Spain         [Barcelona]                    1         False


In [30]:
#Verifica campione di indirizzi UK - .head()
righe_uk = df[df['Hotel_Address'].str.endswith('United Kingdom')]
for idx, indirizzo in righe_uk['Hotel_Address'].drop_duplicates().head(10).items():
    print(f"{idx} --> {indirizzo}")

405 --> 1 15 Templeton Place Earl s Court Kensington and Chelsea London SW5 9NB United Kingdom
971 --> 1 2 Serjeant s Inn Fleet Street City of London London EC4Y 1LL United Kingdom
2008 --> 1 3 Queens Garden Westminster Borough London W2 3BA United Kingdom
3832 --> 1 8 Russell Square Camden London WC1B 5BE United Kingdom
5257 --> 1 Addington Street Lambeth London SE1 7RY United Kingdom
7480 --> 1 Aldwych Westminster Borough London WC2B 4BZ United Kingdom
7961 --> 1 Hamilton Place Park Lane Westminster Borough London W1J 7QY United Kingdom
8301 --> 1 Inverness Terrace Westminster Borough London W2 3JP United Kingdom
11259 --> 1 Kensington Court Kensington and Chelsea London W8 5DL United Kingdom
11345 --> 1 Kings Cross Road Islington London WC1X 9HX United Kingdom


In [31]:
righe_uk = df[df['Hotel_Address'].str.endswith('United Kingdom')]
righe_uk['Hotel_Address'].drop_duplicates().tail(10)

466995           Tilling Road London NW2 1LP United Kingdom
467428    Tower Bridge Road Southwark London SE1 3UN Uni...
503073    Western Avenue Ealing London W5 1HG United Kin...
503486    Western Gateway Newham London E16 1FA United K...
504027    Westminster Bridge Road Lambeth London SE1 7UT...
508196    Westside Wimbledon Common Wimbledon London SW1...
508552    Whitcomb Street Trafalgar Square Westminster B...
509019    Whitehall Place Westminster Borough London SW1...
512545    Wilton Place Knightsbridge Westminster Borough...
512645    Wrights Lane Kensington and Chelsea London W8 ...
Name: Hotel_Address, dtype: str

In [32]:
# Si verifica che "London" sia presente nell'indirizzo UK
righe_uk = df[df['Hotel_Address'].str.endswith('United Kingdom')]
contiene_london = righe_uk['Hotel_Address'].str.contains('London', case=False).sum()

print(f"Righe UK totali: {len(righe_uk)}")
print(f"Righe UK che contengono 'London' nell'indirizzo: {contiene_london}")

Righe UK totali: 262298
Righe UK che contengono 'London' nell'indirizzo: 262298


"London" validato empiricamente sul 100% dei dati di riferimento

In [33]:
# 1. Dizionario Nazione -> Città
mappa_geografica = {
    'United Kingdom': 'London',
    'France': 'Paris',
    'Netherlands': 'Amsterdam',
    'Spain': 'Barcelona',
    'Italy': 'Milan',
    'Austria': 'Vienna'
}

# 2. Funzione per assegnare la città in base al suffisso geografico dell'indirizzo
def assegna_citta(indirizzo):
    indirizzo_str = str(indirizzo).strip()
    for nazione, citta in mappa_geografica.items():
        if indirizzo_str.endswith(nazione):
            return citta
    return 'Unknown'

# 3. Applicazione della funzione per creare la variabile derivata richiesta
df['City'] = df['Hotel_Address'].apply(assegna_citta)

# 4. Verifica della distribuzione finale e controllo assenza di 'Unknown'
print(df['City'].value_counts())
print("\nRighe non riconosciute:", (df['City'] == 'Unknown').sum())

City
London       262298
Barcelona     60149
Paris         59413
Amsterdam     57211
Vienna        38937
Milan         37204
Name: count, dtype: int64

Righe non riconosciute: 0


####
**Gestione missing values — `Reviewer_Nationality`**<br> 
*(richiamo alla sezione 4.1.3.1: qui si completa la gestione missing values anche per `Reviewer_Nationality`, rimandata a questo punto perché la verifica del pattern richiede la colonna `City` appena creata)*


In [34]:
# Verifica `Reviewer_Nationality`
# (richiede la colonna City, creata nella cella precedente)
missing_nat = (df['Reviewer_Nationality'] == '').sum()
print(f"Righe con Reviewer_Nationality mancante: {missing_nat} ({missing_nat/len(df)*100:.2f}%)")

tasso_missing_per_citta = (df[df['Reviewer_Nationality']=='']['City'].value_counts()
                             / df['City'].value_counts() * 100).round(3)
print("\nTasso di missing per città (verifica assenza di pattern sistematico):")
print(tasso_missing_per_citta)

Righe con Reviewer_Nationality mancante: 522 (0.10%)

Tasso di missing per città (verifica assenza di pattern sistematico):
City
Amsterdam    0.121
Barcelona    0.095
London       0.098
Milan        0.118
Paris        0.103
Vienna       0.085
Name: count, dtype: float64


La colonna `Reviewer_Nationality` presenta valori mancanti (522 righe, ~0,10% del dataset).<br>
Il missing risulta distribuito in modo casuale su tutte le città (tasso di missing 0,08%–0,12% ovunque, nessuna concentrazione su hotel specifici) e riguarda un dato anagrafico auto-dichiarato dal recensore, non ricostruibile da altre colonne del dataset.<br>
Si procede per la **conservazione esplicita**: i valori mancanti (stringa vuota `''` dopo la pulizia degli spazi in 4.1.3.0 — nel file originale erano codificati come un singolo carattere spazio `' '`) vengono etichettati come `'Not specified'`, per rendere il missing tracciabile nelle analisi successive senza inventare un valore che non può essere verificato.

In [35]:
# Etichetta esplicita
df['Reviewer_Nationality'] = df['Reviewer_Nationality'].replace('', 'Not specified')

print(f"\nValori mancanti (stringa vuota): {(df['Reviewer_Nationality']=='').sum()}")
print(f"Nazionalità uniche (incluso 'Not specified'): {df['Reviewer_Nationality'].nunique()}")


Valori mancanti (stringa vuota): 0
Nazionalità uniche (incluso 'Not specified'): 227


#####
4.1.3.4.4 Pulizia di days_since_review



Necessaria per il grafico 8 obbligatorio (heatmap correlazioni, punto 4.5.8),<br>che richiede "days since review" tra le variabili numeriche.<br> Nel dataframe originale è una stringa (es. "23 days"), non un numero.

In [36]:
# Verifica formato attuale
print(df['days_since_review'].dtype)
print(df['days_since_review'].head())

str
0     0 days
1     0 days
2     3 days
3     3 days
4    10 days
Name: days_since_review, dtype: str


In [37]:
# Rimuozione del testo "days" / "day" e conversione in numero intero
df['days_since_review'] = (
    df['days_since_review']
    .str.replace(' days', '', regex=False)
    .str.replace(' day', '', regex=False)
    .astype(int)
)

# Verifica
print(df['days_since_review'].dtype)
df['days_since_review'].head()

int64


0     0
1     0
2     3
3     3
4    10
Name: days_since_review, dtype: int64

### 
**4.1.4 Salvataggio del Dataset pulito**

Salviamo il risultato in `DATASET.csv`:<br> sarà il file di partenza sia per il popolamento del database MySQL (Fase 2),<br> sia per l'analisi con pandas nel notebook `02_ANALYSIS_PANDAS_PLOTS.ipynb`.

In [38]:
# Controllo finale del Dataset
print(f"Shape finale del dataset pulito: {df.shape}")
print(f"\nColonne ({len(df.columns)}):")
for col in df.columns:
    print(f"  - {col}")

Shape finale del dataset pulito: (515212, 22)

Colonne (22):
  - Hotel_Address
  - Additional_Number_of_Scoring
  - Review_Date
  - Average_Score
  - Hotel_Name
  - Reviewer_Nationality
  - Negative_Review
  - Review_Total_Negative_Word_Counts
  - Total_Number_of_Reviews
  - Positive_Review
  - Review_Total_Positive_Word_Counts
  - Total_Number_of_Reviews_Reviewer_Has_Given
  - Reviewer_Score
  - Tags
  - days_since_review
  - lat
  - lng
  - Review_Year
  - Review_Month
  - Total_Review_Length
  - Score_Category
  - City


In [39]:
# Salvataggio
df.to_csv('dataset.csv', index=False)
print("Salvato dataset.csv")
print(f"Righe: {len(df)} | Colonne: {len(df.columns)}")

Salvato dataset.csv
Righe: 515212 | Colonne: 22
